### DISCUSSION TITLE
Table locks with enttest

### DISCUSSION THREAD

**Discussion by till** [ent/ent#3474] [Q&A] [answered]
Hi,

I have a small _test suite_ where I have multiple tests assembled.

```go
func TestRequests(t *testing.T) {
 t.Run("first", ...)
 t.Run("second", ...)
}
```

In each test, I setup my database connection and I consistently run into table locks.

```go
	t.Run("first", func(t *testing.T) {
		appName := "rollback-app"
		userName := "user5"

		client := enttest.Open(t, "sqlite3", "file:ent?mode=memory&cache=shared&_fk=1", opts...)
		defer client.Close()
// ...
```

The general setup of each test is that I bootstrap my API (e.g. inject the client) and initially run a request to create a user record and so on. This usually works. Then I run a few queries to manipulate the database to setup scenarios, and then I run another request against my API.

What happens though is, that while the setup works, I get a random error (as in it's always a different test):

```
panic: database table is locked [recovered]
	panic: database table is locked
```

The code is a bit complex, and I haven't yet had a good idea how to isolate it or extract it. But are there best practices or obvious culprits while doing that? I don't want to rule out that I am misunderstanding something either.

I had thought that structuring it like that would also give me the opportunity to run tests in parallel. But apparently, that's not working as I had hoped. Or maybe it's not working with SQLite.

I was also thinking that since I setup the database in each test, it would give me individual databases (and connections), but I am actually not sure if that's true.

The API itself is mostly a couple routes using gorilla/mux which fronts a database (and some container/docker/k8s) stuff for production. In production we use Postgres and don't seem to have these issues.

**Comment by till** [comment, marked as answer]
This seems to work by switching to a file based database (vs. memory based). But I have no idea why. But it reliably works.

**Comment by a8m** [comment]
For the "memory" mode, you can remove the `cache=shared` option to solve this issue (or use a t.Name() for the database name).

**Reply by till** [reply]
Can you combine `t.Name()` with memory mode?

**Comment by fchimpan** [comment]
Me too. 
My settings

```go
client := enttest.Open(t, "sqlite3", "file:"+t.Name()+"?mode=memory&cache=shared&_fk=1")
```
